pkg/iter/entry_iterator.go

```diff
@@ -200,55 +206,85 @@ func (i *mergeEntryIterator) Next() bool {
 	// occurs most often.
+Outer:
 	for i.heap.Len() > 0 {
 		next := i.heap.Peek()
 		entry := next.Entry()
-		if len(i.tuples) > 0 && (i.tuples[0].StreamHash() != next.StreamHash() || !i.tuples[0].Timestamp.Equal(entry.Timestamp)) {
+		if len(i.buffer) > 0 &&
+			(i.buffer[0].streamHash != next.StreamHash() ||
+				!i.buffer[0].Entry.Timestamp.Equal(entry.Timestamp)) {
 			break
 		}
 
 		heap.Pop(i.heap)
-		i.tuples = append(i.tuples, tuple{
-			Entry:         entry,
-			EntryIterator: next,
-		})
-	}
-
-	// shortcut if we have a single tuple.
-	if len(i.tuples) == 1 {
-		i.currEntry.entry = i.tuples[0].Entry
-		i.currEntry.labels = i.tuples[0].Labels()
-		i.currEntry.streamHash = i.tuples[0].StreamHash()
-		i.requeue(i.tuples[0].EntryIterator, false)
-		i.tuples = i.tuples[:0]
-		return true
-	}
-
-	// Find in tuples which entry occurs most often which, due to quorum based
-	// replication, is guaranteed to be the correct next entry.
-	t := i.tuples[0]
-	i.currEntry.entry = t.Entry
-	i.currEntry.labels = t.Labels()
-	i.currEntry.streamHash = i.tuples[0].StreamHash()
-
-	// Requeue the iterators, advancing them if they were consumed.
-	for j := range i.tuples {
-		if i.tuples[j].Line != i.currEntry.entry.Line {
-			i.requeue(i.tuples[j].EntryIterator, true)
-			continue
+		previous := i.buffer
+		var dupe bool
+		for _, t := range previous {
+			if t.Entry.Line == entry.Line {
+				i.stats.AddDuplicates(1)
+				dupe = true
+				break
+			}
+		}
+		if !dupe {
+			i.buffer = append(i.buffer, entryWithLabels{
+				Entry:      entry,
+				labels:     next.Labels(),
+				streamHash: next.StreamHash(),
+			})
 		}
-		// we count as duplicates only if the tuple is not the one (t) used to fill the current entry
-		if i.tuples[j] != t {
-			i.stats.AddDuplicates(1)
+	inner:
+		for {
+			if !next.Next() {
+				continue Outer
+			}
+			entry := next.Entry()
+			if next.StreamHash() != i.buffer[0].streamHash ||
+				!entry.Timestamp.Equal(i.buffer[0].Entry.Timestamp) {
+				break
+			}
+			for _, t := range previous {
+				if t.Entry.Line == entry.Line {
+					i.stats.AddDuplicates(1)
+					continue inner
+				}
+			}
+			i.buffer = append(i.buffer, entryWithLabels{
+				Entry:      entry,
+				labels:     next.Labels(),
+				streamHash: next.StreamHash(),
+			})
 		}
-		i.requeue(i.tuples[j].EntryIterator, false)
+		i.pushBuffer = append(i.pushBuffer, next)
 	}
-	i.tuples = i.tuples[:0]
+
+	for _, ei := range i.pushBuffer {
+		heap.Push(i.heap, ei)
+	}
+	i.pushBuffer = i.pushBuffer[:0]
+
+	i.nextFromBuffer()
+
 	return true
 }
 
+func (i *mergeEntryIterator) nextFromBuffer() {
+	if len(i.buffer) == 1 {
+		i.currEntry.Entry = i.buffer[0].Entry
+		i.currEntry.labels = i.buffer[0].labels
+		i.currEntry.streamHash = i.buffer[0].streamHash
+		i.buffer = i.buffer[:0]
+		return
+	}
+
+	i.currEntry.Entry = i.buffer[0].Entry
+	i.currEntry.labels = i.buffer[0].labels
+	i.currEntry.streamHash = i.buffer[0].streamHash
+	i.buffer = i.buffer[1:]
```

Comment:
Same applies for the mergeSampleIterator

Reply:
Yeah make sense.

Fun fact here ! Technically the if is not required but [:0] improve drastically performance the first element is re-used. I didn't have time to check for [1:] which is the case where  we have many timestamp equal, but I suspect [1:] doesn't allow for reusal of items.

Reply:
I think one way to avoid this would be to save the slice ref on each append and here use

```
i.buffer = append(i.buffer.....

i.savedbuffer = i.buffer

i.buffer = i.savedbuffer[:0]
```